[libpas] Add PAS_ALLOW_UNSAFE_BUFFER_USAGE macro to fix build

zakariaridouh opened the pull request:
#### 2cb447a130b9b26fd5800e940b311feb6613bc5b
<pre>
[libpas] Add PAS_ALLOW_UNSAFE_BUFFER_USAGE macro to fix build
<a href="https://bugs.webkit.org/show_bug.cgi?id=285937">https://bugs.webkit.org/show_bug.cgi?id=285937</a>
<a href="https://rdar.apple.com/142905646">rdar://142905646</a>

Reviewed by Geoffrey Garen.

After recent changes, the compiler is more aggressive
about `unsafe pointer arithmetic [-Werror,-Wunsafe-buffer-usage]`.

We should a macro similar to `WTF_ALLOW_UNSAFE_BUFFER_USAGE` to allow
libpas builds to continue for now, with a followup patch fixing these after.

* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_allocation_result.h:
(pas_allocation_result_zero):
* Source/bmalloc/libpas/src/libpas/pas_compact_cartesian_tree_node_ptr.h:
* Source/bmalloc/libpas/src/libpas/pas_compiler_utils.h: Copied from Source/bmalloc/libpas/src/libpas/pas_compact_cartesian_tree_node_ptr.h.
* Source/bmalloc/libpas/src/libpas/pas_ptr_hash_map.h:

Canonical link: <a href="https://commits.webkit.org/288968@main">https://commits.webkit.org/288968@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eed78a46c2d992b3da167c080da71ce281218ea8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84898 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/4623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/39288 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90043 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35951 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/4713 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/12600 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/66073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23887 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87943 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/3600 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77151 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/46343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/3476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/31378 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35024 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/77863 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/74293 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/32185 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/91416 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/83941 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/12237 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8941 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/74555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/12467 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72963 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73679 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18238 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18060 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/16508 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/4272 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/12189 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/106333 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12024 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25665 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/15518 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
```diff
@@ -3043,7 +3051,6 @@
 				DD4BEC9929CBA49700398E35 /* pas_simple_large_free_heap.c in Sources */,
 				DD4BEDC529CBA49700398E35 /* pas_simple_type.c in Sources */,
 				7234F6B82D01112E000B645D /* pas_small_medium_bootstrap_free_heap.c in Sources */,
-				7234F6C72D011EF3000B645D /* pas_small_medium_bootstrap_free_heap.h in Sources */,
```
Why is this being deleted?

stwrt replied:
Looks like a `.h` got added to the sources 🤔

zakariaridouh replied:
Yeah, not entirely sure why it was reorganized like this.

zakariaridouh committed:
[libpas] Add PAS_ALLOW_UNSAFE_BUFFER_USAGE macro to fix build
https://bugs.webkit.org/show_bug.cgi?id=285937
rdar://142905646

Reviewed by Geoffrey Garen.

After recent changes, the compiler is more aggressive
about `unsafe pointer arithmetic [-Werror,-Wunsafe-buffer-usage]`.

We should a macro similar to `WTF_ALLOW_UNSAFE_BUFFER_USAGE` to allow
libpas builds to continue for now, with a followup patch fixing these after.

* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/libpas.xcodeproj/project.pbxproj:
* Source/bmalloc/libpas/src/libpas/pas_allocation_result.h:
(pas_allocation_result_zero):
* Source/bmalloc/libpas/src/libpas/pas_compact_cartesian_tree_node_ptr.h:
* Source/bmalloc/libpas/src/libpas/pas_compiler_utils.h: Copied from Source/bmalloc/libpas/src/libpas/pas_compact_cartesian_tree_node_ptr.h.
* Source/bmalloc/libpas/src/libpas/pas_ptr_hash_map.h:

Canonical link: https://commits.webkit.org/288968@main
